yt_dlp/extractor/telegram.py

```diff
@@ -1,41 +1,141 @@
+        },
+    }, {
+        # 2-video post with --no-playlist
+        'url': 'https://t.me/vorposte/29343',
+        'md5': '1724e96053c18e788c8464038876e245',
+        'info_dict': {
+            'id': '29343',
+            'ext': 'mp4',
+            'title': 'Форпост',
+            'uploader': 'vorposte',
+            'thumbnail': r're:^https?://.+',
+            'timestamp': 1666384480,
+            'upload_date': '20221021',
+            'duration': 35,
+        },
+        'params': {
+            'noplaylist': True,
+        }
+    }, {
+        # 2-video post with 'single' query param
+        'url': 'https://t.me/vorposte/29342?single',
+        'md5': 'd20b202f1e41400a9f43201428add18f',
+        'info_dict': {
+            'id': '29342',
+            'ext': 'mp4',
+            'title': 'Форпост',
+            'description': 'md5:4d573fd641c3a21f58ad6bb5f16fe526',
+            'uploader': 'vorposte',
+            'thumbnail': r're:^https?://.+',
+            'timestamp': 1666384480,
+            'upload_date': '20221021',
+            'duration': 33,
         },
+    }, {
+        # 2-video post with 'single' query param and --yes-playlist
+        'url': 'https://t.me/vorposte/29343?single',
+        'info_dict': {
+            'id': 'vorposte-29343',
+            'title': 'Форпост 29343',
+            'description': 'md5:4d573fd641c3a21f58ad6bb5f16fe526',
+        },
+        'playlist_count': 2,
+        'params': {
+            'noplaylist': False,
+            'skip_download': True,
+        }
     }]
 
     def _real_extract(self, url):
-        video_id = self._match_id(url)
-        webpage = self._download_webpage(url, video_id, query={'embed': 0})
-        webpage_embed = self._download_webpage(url, video_id, query={'embed': 1}, note='Downloading ermbed page')
+        uploader, msg_id = self._match_valid_url(url).group('uploader', 'id')
 
-        formats = [{
-            'url': self._proto_relative_url(self._search_regex(
-                '<video[^>]+src="([^"]+)"', webpage_embed, 'source')),
-            'ext': 'mp4',
-        }]
-        self._sort_formats(formats)
-
-        return {
-            'id': video_id,
-            'title': self._html_search_meta(['og:title', 'twitter:title'], webpage, default=None),
-            'description': self._html_search_meta(
-                ['og:description', 'twitter:description'], webpage,
-                default=clean_html(get_element_by_class('tgme_widget_message_text', webpage_embed))),
-            'thumbnail': self._search_regex(
-                r'tgme_widget_message_video_thumb"[^>]+background-image:url\(\'([^\']+)\'\)',
-                webpage_embed, 'thumbnail'),
-            'formats': formats,
+        query = {'embed': '1'}
+        if self.get_param('noplaylist'):
+            query['single'] = True
+        elif self.get_param('noplaylist') is False:
+            pass
+        elif '?single' in url:  # param 'noplaylist' is None
+            query['single'] = True
```

Comment:
Use utils.parse_qs and self._yes_playlist

noplaylist=None is not a thing. If `?single` is given, that should take precedence over yes/no-playlist

Reply:
`parse_qs` cannot parse a query param without a value, it returns `None`

Reply:
Do:
```diff
diff --git a/yt_dlp/utils.py b/yt_dlp/utils.py
index de91c66f1..0b1095f55 100644
--- a/yt_dlp/utils.py
+++ b/yt_dlp/utils.py
@@ -3067,8 +3067,8 @@ def escape_url(url):
     ).geturl()


-def parse_qs(url):
-    return urllib.parse.parse_qs(urllib.parse.urlparse(url).query)
+def parse_qs(url, **kwargs):
+    return urllib.parse.parse_qs(urllib.parse.urlparse(url).query, **kwargs)


 def read_batch_urls(batch_fd):
```
and then `parse_qs(url, keep_blank_values=True)`

Reply:
done... I *think* I've managed `?single` precedence and `_yes_playlist` output correctly